tests/engine/training/test_components.py

```diff
@@ -23,3 +34,154 @@ def test_cached_component_returns_value_from_cache(default_model_storage: ModelS
 
     returned_output = node()["cached"]
     assert returned_output.text == "Cache me!!"
+
+
+def test_cached_component_replace_schema_node():
+    schema_node = SchemaNode(
+        needs={"i1": "first_input", "i2": "second_input"},
+        uses=FingerprintComponent,
+        fn="add",
+        constructor_name="load",
+        config={"a": 1},
+        eager=False,
+        is_input=False,
+        resource=Resource("hello"),
+    )
+
+    CachedComponent.replace_schema_node(schema_node, 2)
+
+    assert schema_node == SchemaNode(
+        needs={"i1": "first_input", "i2": "second_input"},
+        uses=CachedComponent,
+        fn="get_cached_output",
+        constructor_name="create",
+        config={"output": 2},
+        eager=False,
+        is_input=False,
+        resource=Resource("hello"),
+    )
+
+
+def test_fingerprint_component_replace_schema_node(temp_cache: TrainingCache):
+    schema_node = SchemaNode(
+        needs={"i1": "first_input", "i2": "second_input"},
+        uses=CachedComponent,
+        fn="add",
+        constructor_name="load",
+        config={"a": 1},
+        eager=False,
+        is_input=False,
+        resource=Resource("hello"),
+    )
+
+    FingerprintComponent.replace_schema_node(schema_node, temp_cache, "some_node_name")
+
+    assert schema_node == SchemaNode(
+        needs={"i1": "first_input", "i2": "second_input"},
+        uses=FingerprintComponent,
+        fn="run",
+        constructor_name="create",
+        config={"a": 1, "cache": temp_cache, "node_name": "some_node_name"},
+        eager=True,
+        is_input=False,
+        resource=Resource("hello"),
+    )
+
+
+@dataclasses.dataclass
+class FingerprintableText:
+    text: Text
+
+    def fingerprint(self):
+        return self.text
+
+
+def test_fingerprint_component_hit(
+    default_model_storage: ModelStorage, temp_cache: TrainingCache
+):
+
+    cached_output = CacheableText("Cache me!!")
+    output_fingerprint = uuid.uuid4().hex
+    component_config = {"x": 1}
+    fingerprint_key = fingerprinting.calculate_fingerprint_key(
+        node_name="original_node",
+        config=component_config,
+        inputs={
+            "param_1": FingerprintableText("input_1"),
+            "param_2": FingerprintableText("input_2"),
+        },
+    )
+    temp_cache.cache_output(
+        fingerprint_key=fingerprint_key,
+        output=cached_output,
+        output_fingerprint=output_fingerprint,
+        model_storage=default_model_storage,
+    )
```

Comment:
how about adding comments to the tests to make it easier to understand what we are doing? E.g.
```suggestion
    cached_output = CacheableText("Cache me!!")
    output_fingerprint = uuid.uuid4().hex
    component_config = {"x": 1}
    
    # Pretend the output for the replaced component was cached in a previous run
    fingerprint_key = fingerprinting.calculate_fingerprint_key(
        node_name="original_node",
        config=component_config,
        inputs={
            "param_1": FingerprintableText("input_1"),
            "param_2": FingerprintableText("input_2"),
        },
    )
    temp_cache.cache_output(
        fingerprint_key=fingerprint_key,
        output=cached_output,
        output_fingerprint=output_fingerprint,
        model_storage=default_model_storage,
    )
```